cpp/src/arrow/engine/substrait/extension_set.cc

```diff
@@ -174,31 +134,51 @@ Result<ExtensionSet> ExtensionSet::Make(std::vector<util::string_view> uris,
 }
 
 Result<ExtensionSet::TypeRecord> ExtensionSet::DecodeType(uint32_t anchor) const {
-  if (anchor >= types_.size() || types_[anchor].id.empty()) {
+  if (types_.find(anchor) == types_.end() || types_.at(anchor).id.empty()) {
     return Status::Invalid("User defined type reference ", anchor,
                            " did not have a corresponding anchor in the extension set");
   }
-  return types_[anchor];
+  return types_.at(anchor);
 }
 
 Result<uint32_t> ExtensionSet::EncodeType(const DataType& type) {
   if (auto rec = registry_->GetType(type)) {
-    return impl_->EncodeType(*rec, this);
+    RETURN_NOT_OK(this->AddUri(rec->id));
+    auto it_success =
+        types_map_.emplace(rec->id, static_cast<uint32_t>(types_map_.size()));
+    if (it_success.second) {
+      if (types_.find(static_cast<unsigned int>(types_.size())) != types_.end()) {
+        return Status::Invalid("Key already exist in the uris map");
+      }
```

Comment:
```suggestion
      DCHECK_EQ(types_.find(static_cast<unsigned int>(types_.size())), types_.end()) << "Type existed in types_ but not types_map_.  ExtensionSet is inconsistent";
```
There really isn't much a user can do with this message.  If we get here I think it would have to be a bug (possibly reusing an ExtensionSet) so we might as well make it a `DCHECK`.

Reply:
Using `DCHECK` now instead of only returning a Invalid status